concordium-consensus/src/Concordium/KonsensusV1/Consensus/Blocks.hs

```diff
@@ -0,0 +1,1388 @@
+      MonadState (SkovData (MPV m)) m,
+      BlockStateStorage m,
+      BlockState m ~ HashedPersistentBlockState (MPV m),
+      IsConsensusV1 (MPV m),
+      MonadLogger m
+    ) =>
+    m (Maybe (BakeBlockInputs (MPV m)))
+prepareBakeBlockInputs = runMaybeT $ do
+    canBake <- roundStatus . rsRoundEligibleToBake <<.= False
+    -- Terminate if we've already tried to bake for this round.
+    guard canBake
+    sd <- get
+    let rs = sd ^. roundStatus
+    -- Note: we rely on the current round being always in sync with the previous round TC/
+    -- highest QC.
+    let bbiRound = rs ^. rsCurrentRound
+    -- Check that we haven't baked for this or a more recent round.
+    guard (bbiRound > sd ^. persistentRoundStatus . prsLastBakedRound)
+    -- Terminate if we do not have baker credentials.
+    bbiBakerIdentity@BakerIdentity{..} <- MaybeT (view Consensus.bakerIdentity)
+    -- Determine the parent block and its quorum certificate, as well as the timeout certificate if
+    -- applicable.
+    let ( bbiTimeoutCertificate,
+          CertifiedBlock
+            { cbQuorumBlock = bbiParent,
+              cbQuorumCertificate = bbiQuorumCertificate
+            }
+            )
+                | Present RoundTimeout{..} <- rs ^. rsPreviousRoundTimeout =
+                    ( Present rtTimeoutCertificate,
+                      if cbEpoch highestCB > cbEpoch rtCertifiedBlock
+                        then rtCertifiedBlock
+                        else highestCB
+                    )
+                | otherwise =
+                    ( Absent,
+                      highestCB
+                    )
+              where
+                highestCB = rs ^. rsHighestCertifiedBlock
+    let bbiEpoch = sd ^. currentEpoch
+    let bbiEpochFinalizationEntry
+            | bbiEpoch > qcEpoch bbiQuorumCertificate =
+                -- This assertion should not fail because the invariant on
+                -- '_lastEpochFinalizationEntry' requires the entry to be present whenever
+                -- @_currentEpoch > blockEpoch _lastFinalized@, and @bbiQuorumCertificate@ is for
+                -- the parent block which is in at least the epoch of the last finalized block.
+                assert (isPresent finEntry) finEntry
+            | otherwise = Absent
+          where
+            finEntry = sd ^. lastEpochFinalizationEntry
+    bbiEpochBakers <-
+        if isAbsent bbiEpochFinalizationEntry
+            then getCurrentEpochBakers (bpState bbiParent)
+            else getNextEpochBakers (bpState bbiParent)
+    parentSeedState <- getSeedState (bpState bbiParent)
+    let bbiLeadershipElectionNonce =
+            if isAbsent bbiEpochFinalizationEntry
+                then parentSeedState ^. currentLeadershipElectionNonce
+                else nonceForNewEpoch bbiEpochBakers parentSeedState
+    let leader = getLeaderFullBakers bbiEpochBakers bbiLeadershipElectionNonce bbiRound
+    guard (leader ^. bakerIdentity == bakerId)
```

Comment:
We're not really making the block here - i.e. executing the transactions. 

This approach is chosen for simplicity as we're just looking it up from the block state (and we usually don't want to have multiple copies of the same data around as this just makes invariants more complex.

I don't think it will make much sense now to make the code more complex. But I'll add an entry to the optimizations issue so we can evaluate later on.